File: src/toolchains/docker.py
import subprocess
import json
import shutil


def scan():
    result = {
        "tool": "docker",
        "available": False,
        "version": None,
        "disk_usage": {},
        "issues": [],
        "status": "ok",
    }

    if not shutil.which("docker"):
        result["status"] = "skipped"
        result["issues"].append("Docker not installed")
        return result

    try:
        d_v = subprocess.run(
            ["docker", "version", "--format", "{{.Server.Version}}"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        result["version"] = d_v.stdout.strip() or "unknown"

        info = subprocess.run(
            ["docker", "info"], capture_output=True, text=True, timeout=10
        )
        if info.returncode != 0:
            result["status"] = "error"
            result["issues"].append("Docker daemon not running")
            return result

        df = subprocess.run(
            ["docker", "system", "df", "--format", "json"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        if df.returncode == 0 and df.stdout.strip():
            for line in df.stdout.strip().split("\n"):
                if line.strip():
                    try:
                        data = json.loads(line)
                        typ = data.get("Type", "")
                        result["disk_usage"][typ] = {
                            "total": data.get("TotalCount", "?"),
                            "size": data.get("Size", "?"),
                            "reclaimable": data.get("ReclaimableSize", "?"),
                        }
                    except json.JSONDecodeError:
                        pass

        images = subprocess.run(
            ["docker", "images", "--filter", "dangling=true", "-q"],
            capture_output=True,
            text=True,
            timeout=10,
        )
        dangling = [x for x in images.stdout.strip().split("\n") if x]
        if dangling:
            result["issues"].append(f"{len(dangling)} dangling image(s)")

        result["status"] = "ok"
    except Exception as e:
        result["status"] = "error"
        result["issues"].append(str(e))

    return result

File: src/toolchains/docker.rs
use super::*;

pub async fn scan() -> ScanResult {
    if !which("docker") {
        return ScanResult::skipped("Docker not installed");
    }

    let mut result = ScanResult::new("docker");

    let info_cmd = Command::new("docker")
        .args(["info", "--format", "{{json .}}"])
        .output();

    let (ver_res, info_res) = tokio::join!(
        run_cmd("docker", &["--version"], None),
        tokio::time::timeout(Duration::from_secs(10), info_cmd)
    );

    if let Ok(ver) = ver_res {
        result.version = Some(ver);
    }

    let info_check = match info_res {
        Ok(Ok(out)) => out,
        _ => {
            result.status = "error".into();
            result.issues.push("Docker daemon not running".into());
            return result;
        }
    };

    if info_check.status.success() {
        if let Ok(info) =
            serde_json::from_str::<serde_json::Value>(&String::from_utf8_lossy(&info_check.stdout))
        {
            let mut disk = serde_json::Map::new();
            if let Some(driver) = info["Driver"].as_str() {
                disk.insert("driver".into(), serde_json::Value::String(driver.into()));
            }
            result.disk_usage = Some(serde_json::Value::Object(disk));
        }
    }

    result
}
